README.md
`neotoma`
========

[![Build Status](https://api.travis-ci.org/ropensci/neotoma.png)](https://travis-ci.org/ropensci/neotoma)
[![Build status](https://ci.appveyor.com/api/projects/status/t2xyqbs0d8h998cb/branch/master)](https://ci.appveyor.com/project/sckott/neotoma/branch/master)
[![codecov.io](https://codecov.io/github/ropensci/neotoma/coverage.svg?branch=master)](https://codecov.io/github/ropensci/neotoma?branch=master)
[![rstudio mirror downloads](http://cranlogs.r-pkg.org/badges/grand-total/neotoma)](https://github.com/metacran/cranlogs.app)
[![cran version](http://www.r-pkg.org/badges/version/neotoma)](https://cran.r-project.org/package=neotoma)

The `neotoma` package is a programmatic R interface to the [Neotoma Paleoecological Database](http://www.neotomadb.org/). The package is intended to both allow users to search for sites and to download data for use in analyical workflows of paleoecological research.

`neotoma` is part of the [rOpenSci](http://ropensci.org) project and is also hosted on [Figshare](http://dx.doi.org/10.6084/m9.figshare.677131).  The `neotoma` package is also available on [CRAN](https://cran.r-project.org/package=neotoma) as of May 3, 2015.

For more information on the package please refer to: 

Goring, S., Dawson, A., Simpson, G. L., Ram, K., Graham, R. W., Grimm, E. C., & Williams, J. W.. (2015). neotoma: A Programmatic Interface to the Neotoma Paleoecological Database. *Open Quaternary*, 1(1), Art. 2. DOI: [10.5334/oq.ab](http://doi.org/10.5334/oq.ab)

For ongoing news, issues or information please join the [Neotoma Slack server](https://bit.ly/2FrZyYD) (and add the #r channel).  [Slack](http://www.slack.com/) is a real time chat application and collaboration hub with mobile and desktop applications.

### Development

*We welcome contributions from any individual, whether code, documentation, or issue tracking.  All participants are expected to follow the [code of conduct](https://github.com/ROpensci/neotoma/blob/master/code_of_conduct.md) for this project.*

+ [Simon Goring](http://goring.org) - University of Wisconsin-Madison, Department of Geography

### Contributors
+ [Gavin Simpson](http://www.fromthebottomoftheheap.net/) - University of Regina, Department of Biology
+ [Jeremiah Marsicek](http://geoweb.uwyo.edu/ggstudent/jmarsice/Site/Home.html) - University of Wyoming, Department of Geology and Geophysics
+ [Karthik Ram](http://nature.berkeley.edu/~kram/) - University of California - Berkely, Berkeley Institue for Data Science.
+ [Luke Sosalla](https://github.com/sosalla) - University of Wisconsin, Department of Geography

Package functions resolve various Neotoma APIs and re-form the data returned by the Neotoma database into R data objects.  The format of the Neotoma data, and the actual API functions can be accessed on the Neotoma API [website](http://api.neotomadb.org/doc/resources/home).

If you have used the package please consider providing us feedback through a [short survey](https://docs.google.com/forms/d/e/1FAIpQLSdRNat6L9grRF0xU5gibkr26xq9jD9wyHgw_AWxhrgn0lWv7w/viewform).

### Install `neotoma`

+ CRAN:
```r
install.packages('neotoma')
```

+ Development version from GitHub:
```r
install.packages("devtools")
library(devtools)
install_github("ropensci/neotoma")
library(neotoma)
```

### Currently implemented in `neotoma`

More functions are available through the package help.  These represent the core functions:

+ `get_site` - obtain information on sites in the Neotoma dataset (which may contain multiple datasets). [API](http://api.neotomadb.org/doc/resources/sites)
+ `get_dataset` - obtain dataset metadata from Neotoma. [API](http://api.neotomadb.org/doc/resources/datasets)
+ `get_download` - obtain full datasets (pollen or mammal) from Neotoma. [API](http://api.neotomadb.org/doc/resources/downloads)
+ `compile_list` - using established pollen-related taxonomies from the literature, take the published taxon list and standardize it to allow cross site analysis.
+ `get_contact` - find contact information for data contributors to Neotoma. [API](http://api.neotomadb.org/doc/resources/contacts)
+ `get_publication` - obtain publication information from Neotoma. [API](http://api.neotomadb.org/doc/resources/publications)
+ `get_table` - return matrices corresponding to one of the Neotoma database tables. [tables](http://api.neotomadb.org/doc/resources/dbtables)
+ `get_taxa` - Get taxon information from Neotoma. [API](http://api.neotomadb.org/doc/resources/taxa)
+ `get_chroncontrol` - Get chronological information used to build the age-depth model for the record. [API](http://api.neotomadb.org/doc/resources/chroncontrol)

### Recent Changes
+1.7.6: Moved the `pollen.equiv` pollen equivalence file into an rda file, can be accessed using `data("pollen.equiv")`; Changed the way that geographic identifiers are used and converted in `param_check` removes a bug where place names that are repeated were getting flagged multiple times; fixed units issue in `get_closest()`.
+ 1.7.5: Introduction of the `sf` package into the package.  Using `sf` to buffer and transform coordinates to allow a proper buffer and bounding box to be generated to identify the *n* closest sites to a point of interest; Added a warning to `plot_leaflet()` for the default response; Updated `param_check()` to include new dataset types within the database.
+ 1.7.4: Bug fix: `get_dataset(gpid=123)` was returning an error, fix corrects the error to allow unassigned `x` variables.  Updated the allowable dataset types for searching to reflect the larger set of dataset types within Neotoma.
+ 1.7.3: Added numeric/integer methods to the `get_site()` and `get_dataset()` functions so that a vector of dataset or siteids can be passed to improve more general workflow methods.
+ 1.7.2: Bugfixes, added the `taxa()` function to easily extract taxa from one or multiple download objects.
+ 1.7.1: Bugfix for `compile_download()`, single sample downloads were failing to compile properly, added the `taxa()` function to extract taxa lists from large download objects.
+ 1.7.0: Added `plot_leaflet()` to allow interactive exploration of downloaded Neotoma data.  Integrates with the Neotoma Explorer.  Minor bugfix for `get_download()` to allow records to be sent to Neotoma and to be filtered.
+ 1.6.2: Improved the basic `plot()` method based on tests against Tilia files in the Neotoma Holding Tank & built more robust interpolation in `read_bacon()` so that age models without interpolated dates can still be imported. `browse()` now opens multiple datastes in the Neotoma Explorer at once.
+ 1.6.1: New `Stratiplot()` method, using the `analogue` package to plot dataset diagrams from `download` and `download_list` objects, bug fixes for `write_agefile()` and a new function, `read_bacon()`, to read in and integrate Bacon chronologies into `download` objects.
+ 1.6.0: Support for vector inputs in the `gpid` selection. Added a `get_closest()` function to find the closest sample site. Mostly clean-up of reported bugs by users. Revised examples for faster check speed.
+ 1.5.1: Minor fix to the `get_dataset()` for site level data to account for some datasets with empty submission data.  Some style changes to code (non-functional changes)
+ 1.5.0: More extensive testing to support multiple dataset types.  Water chemistry datasets still unsupported. Function `read.tilia()` added to read Tilia (http://tiliait.com) style XML files. Moved to using `xml2`, `httr` and `jsonlite` to support parsing.
+ 1.4.1: Small changes to `get_geochron()` to address bug reports and improve object `print` methods.
+ 1.4.0: Added `plot()` method for datasets, sites & downloads.  Fixed a bug with records missing chronologies.

### A few examples:

#### Find the distribution of sites with Mammoth fossils in Neotoma

```r
#  Example requires the mapdata package:
library('mapdata')

#  You may use either '%' or '*' as wildcards for search terms:
test <- get_dataset(taxonname='Mammuthus*')

The API call was successful, you have returned  3273 records.

site.locs <- get_site(test)

# A crude way of making the oceans blue.
plot(1, type = 'n',
     xlim=range(site.locs$long)+c(-10, 10),
     ylim=range(site.locs$lat)+c(-10, 10),
     xlab='Longitude', ylab = 'Latitude')
rect(par("usr")[1],par("usr")[3],par("usr")[2],par("usr")[4],col = "lightblue")
map('world',
    interior=TRUE,
    fill=TRUE,
    col='gray',
    xlim=range(site.locs$long)+c(-10, 10),
    ylim=range(site.locs$lat)+c(-10, 10),
    add=TRUE)

points(site.locs$long, site.locs$lat, pch=19, cex=0.5, col='red')

```
![thing](inst/img/mammothsites.png)

#### Plot the proportion of publications per year for datasets in Neotoma

```R
# Requires ggplot2
library('ggplot2')
library('plyr')
pubs <- get_publication()

pub.years <- ldply(pubs, "[[", "meta")

ggplot(data=pub.years, aes(x = year)) +
     stat_bin(aes(y=..density..*100, position='dodge'), binwidth=1) +
     theme_bw() +
     ylab('Percent of Publications') +
     xlab('Year of Publication') +
     scale_y_continuous(expand = c(0, 0.1)) +
     scale_x_continuous(breaks = seq(min(pub.years$year, na.rm=TRUE), 2014, by=20))

```

![thing](inst/img/histogramplot.png)

#### Cumulative plot of record uploads to Neotoma since 1998.

Found at [this gist](https://gist.github.com/SimonGoring/718a654f304f2d16ce4b)

![cumulativerecords](https://cloud.githubusercontent.com/assets/1619126/9884174/0026b83a-5ba4-11e5-9f1a-4a6874ceceb6.png)

#### Obtain records & Rebuild Chronologies with Bacon

Found at [this gist](https://gist.github.com/SimonGoring/877dd71cc3ad6bf8531e).  Prepared in part for a Bacon (Blaauw & Christen, 2011) workshop at the 2015 International Limnogeology Conference in Reno-Tahoe, Nevada led by Amy Myrbo (University of Minnesota).

#### Simple paleo-data visualization

Simple paleo-data visualization in R, linking the `rioja`, `neotoma` and `dplyr` packages.  Found at [this gist](https://gist.github.com/SimonGoring/dbb4c8e0087882dc143baa89fa041d2b).

![gif](inst/img/inkspot_neotoma.gif)

#### Find all site elevations in California:

Found at [Simon Goring's gist.](https://gist.github.com/SimonGoring/6a2ba1d55a3a7f78723b37e896b55b70).

#### Match all Neotoma taxa to external databases using `taxize`:

Found at [Simon Goring's gist.](https://gist.github.com/SimonGoring/24fb1228204f768f3f0020f37060db18).

#### Other Resources Using `neotoma`

*  [A simple `neotoma` workshop](http://www.goring.org/resources/Neotoma_Lesson.html)
*  [Data exploration and chronologies with `neotoma`](http://www.goring.org/resources/neotoma_lab_meeting.html)

### `neotoma` Workshops

We have provided a set of educational tools through the [NeotomaDB](http://github.com/neotomaDB) GitHub repository in the [Workshops](http://github.com/NeotomaDB/Workshops) repository.  These are free to share, and can be modified as needed.


NEWS
neotoma 1.7.6
====================
o Moved the `pollen.equiv` pollen equivalence file into an rda file, can be accessed using `data("pollen.equiv")`
o Changed the way that geographic identifiers are used and converted in `param_check` removes a bug where place names that are repeated were getting flagged multiple times.

neotoma 1.7.5
====================
o Introduction of the `sf` package into the package.  Using `sf` to buffer and transform coordinates to allow a proper buffer and bounding box to be generated to identify the *n* closest sites to a point of interest.
o Added a warning to `plot_leaflet()` for the default response.
o Updated `param_check()` to include new dataset types within the database.

neotoma 1.7.4
====================
o Bug fix: `get_dataset(gpid=123) was returning an error, fix corrects the error to allow unassigned `x` variables.
o Updated the allowable dataset types for searching to reflect the larger set of dataset types within Neotoma.

neotoma 1.7.3
====================
o Added numeric/integer methods to the `get_site()` and `get_dataset()` functions so that a vector of dataset or siteids can be passed to improve more general workflow methods.

neotoma 1.7.2
====================
o Bugfixes, added the `taxa()` function to easily extract taxa from one or multiple download objects.

neotoma 1.7.1
====================
o Bugfix for `compile_download()`, single sample downloads were failing to compile properly.

neotoma 1.7.0
====================
o Added `plot_leaflet()` to allow interactive exploration of downloaded Neotoma data.  Integrates with the Neotoma Explorer.
o Minor bugfix for `get_download` to allow records to be sent to Neotoma and to be filtered.

neotoma 1.6.2
====================
o Bugfix in `get_chroncontrol()`, `read.tilia()` based on experience in extensive testing.  Improved the basic `plot` method based on tests against Tilia files in the Neotoma Holding Tank.
o Built more robust interpolation in `read_bacon` so that age models without interpolated dates can still be imported.
o `browse` now opens multiple datastes in the Neotoma Explorer at once.

neotoma 1.6.1
====================
o Bugfix
o New `Stratiplot` method, using the `analogue` package to plot dataset diagrams from `download` and `download_list` objects.
o Bug fixes for `write_agefile`
o New function `read_bacon` to read in and integrate Bacon chronologies into `download` objects.

neotoma 1.6.0
====================
o Support for vector inputs in the `gpid` selection.
o Added a `get_closest` function to find the closest datasets to a sample site.
o Mostly clean-up of reported bugs by users.
o Revised examples for faster check speed.

neotoma 1.5.1
====================
o Fix for datasets with empty submission dates.
o Some minor (non-functional) code cleaning.
o Added extra tests.

neotoma 1.5.0
====================
o More extensive testing to support multiple dataset types.  Water chemistry datasets still unsupported.
o Function `read.tilia` added to read Tilia (http://tiliait.com) style XML files.
o Moved to using `xml2`, `httr` and `jsonlite` to support parsing.

neotoma 1.4.0
====================
o The few records without chronologies were causing errors when `get_download` was applied.
o Added plot method.

neotoma 1.3.3
====================
o Critical bug fix: `get_download` now correctly assigns the default chronology to `sample.meta`.  All other chronologies (including the default) are stored in the `chronologies` list for the `download`.
o Bugfix for issue #187, empty `dataset_list`s now return a NULL value, rather than breaking the `print` method.

neotoma 1.3.2
====================
o Bugfix for `get_download`, assignment to the `dataset.id` was incorrectly placed causing the columns to be out of place relative to the data.

neotoma 1.3.1
====================
o Bugfix for `write_agefile`
o Added a `browse` method for datasets, that will open a browser window for a specific dataset or download.
o The deprecated `compile_list` function has now been removed.  Use `compile_taxa` instead.

neotoma 1.3.0
====================
o Improved functionality of get_chroncontrol to allow it to deal with empty tables and work directly with download and download_list objects.

neotoma 1.2-0
====================
o Added new datasettypes for records, including water chemistry, charcoal, XRD and others.

neotoma 1.1-0
====================
o Publication of Goring, S., Dawson, A., Simpson, G. L., Ram, K., Graham, R. W., Grimm, E. C., & Williams, J. W.. (2015). neotoma: A Programmatic Interface to the Neotoma Paleoecological Database, 1(1), Art. 2. DOI: http://doi.org/10.5334/oq.ab

neotoma 0.1
====================
o continued development of API implementation and working toward a common 
  standard for data output.

o Continued development of tests for the package.

o Changed function names to singular.

R/get_closest.R
#' Find the closest dataset records to a site, dataset or long/lat pair in Neotoma
#'
#' Passing in a download object the function outputs a Bacon or Clam formatted file to a
#' user defined destination for age modelling with existing age-depth modeling software.
#'
#' @importFrom utils write.csv write.table
#' @importFrom assertthat assert_that
#' @import sf
#' @param x A vector long/lat pair, or a dataset, site or download.
#' @param n The maximum number of records to return (in the case of ties the return may be larger)
#' @param buffer The size of the buffer for dataset search (in kilometers)
#' @param ... optional arguments to pass into \code{get_dataset}.
#' 
#' @details The function uses the \code{sf} package to generate a circular buffer around a point of interest.  
#' From there a square bounding box is sent to Neotoma using the \code{get_dataset()} function.  To use the buffering 
#' function we must convert from long/lat to UTM coordinates, which we do by guessing the UTM zone of the point of interest.
#' Details can be found in the function's R code hosted on GitHub: \url{https://github.com/ropensci/neotoma/blob/master/R/get_closest.R}
#' @author Simon J. Goring \email{simon.j.goring@@gmail.com}, Andria Dawson \email{andria.dawson@gmail.com}
#' 
#' @return This command returns a \code{dataset} or \code{dataset_list}, or NULL if no records exist within the bounding box.
#' 
#' @examples \dontrun{
#' #  The point of pulling chronology tables is to re-build or examine the chronological 
#' #  information that was used to build the age-depth model for the core.
#' # Find the closest records to Madison, WI:
#' get_closest(x = c(-89.4012, 43.0731), n = 10, buffer = 5000, datasettype = "pollen")
#' }
#' @references
#' Neotoma Project Website: http://www.neotomadb.org
#' API Reference:  http://api.neotomadb.org/doc/resources/contacts
#' @keywords Neotoma Palaeoecology API
#' @export

get_closest <- function(x, n, buffer, ...) {
  UseMethod('get_closest')
}

#' @export
get_closest.default <- function(x, n = 5, buffer = 100, ...) {
  
  assertthat::assert_that(length(x) == 2, 
                          msg = "You must pass an x, y object (e.g., c(10, -20)).")
  assertthat::assert_that(x[1] >= -180 & x[1] <= 180, 
                          msg = "You must pass an array of long/lat with long values from -180 - 180.")
  assertthat::assert_that(x[2] >= -90 & x[2] <= 90, 
                          msg = "You must pass an array of long/lat with lat values from -90 - 90.")

  assertthat::assert_that(buffer < (40075 / 2),
                          msg = paste0("The circumference of the earth is 40000km. You provided a buffer diameter of ", 
                                       buffer, 
                                       "km.  Maybe you entered the value in meters?"))
  utm_w <- seq(-180, 180, by = 6)
  
  proj <- paste0("+proj=utm +zone=",findInterval(x[1], utm_w))
  
  if (x[2] < 0) proj <- paste0(proj, " +south")
  
  bbox <- sf::st_point(x) %>%
    sf::st_sfc() %>%
    sf::st_set_crs(4326) %>% 
    sf::st_transform(crs = proj) %>%
    sf::st_buffer(buffer * 1000) %>%
    sf::st_transform(4326) %>%
    sf::st_bbox() %>% 
    as.numeric()
    
  buff_sets <- suppressMessages(get_dataset(loc = bbox, ...))
  
  if (is.null(buff_sets)) {
    message(paste0("There are no sites within ", buffer, "km of coordinates [", x[1], ", ", x[2], "]"))
    return(NULL)
  } else {
    sites <- get_site(buff_sets)
  }
  
  # From: http://www.r-bloggers.com/r-subsetting-a-list-based-on-a-condition/
  deg2rad <- function(deg) return(deg*pi/180)
  gcd.slc <- function(long1, lat1, long2, lat2) {
    R <- 6371000 # Earth mean radius [km]
    d <- acos(sin(deg2rad(lat1))*sin(deg2rad(lat2)) + cos(deg2rad(lat1)) * 
                cos(deg2rad(lat2)) * cos(deg2rad(long2) - deg2rad(long1))) * R
    return(d) # Distance in m
  }
  
  dists <- sapply(1:nrow(sites), 
                  function(i) gcd.slc(x[1], x[2], sites$long[i], sites$lat[i]) / 1000)
  
  if (all(dists > buffer)) {
    message(paste0("There are no sites within ", buffer, "km of coordinates [", x[1], ", ", x[2], "]"))
    return(NULL)
  }
  
  message(paste0("The API call returns ", sum(dists < buffer, na.rm = TRUE),
          " records within the ", buffer, "km buffer."))
  
  if (length(which(dists < buffer)) <= n) {
    data_out <- buff_sets[which(dists < buffer)]
  } else {
    set <- which((rank(dists, ties.method = 'min') <= n) & (dists < buffer))
    message("Returning the closest ", n, " records.")
    data_out <- buff_sets[set]
  }
  
  if (length(data_out) == 1) {
    class(data_out) <- c("dataset", "list")
  } else {
    class(data_out) <- c("dataset_list", "list")
  }
  
  return(data_out)

}

#' @export
get_closest.site <- function(x, n, buffer, ...) {
  coords <- c(x$long, x$lat)
  get_closest(coords, n, buffer, ...)
}

#' @export
get_closest.dataset <- function(x, n, buffer, ...) {
  coords <- c(get_site(x)$long, get_site(x)$lat)
  get_closest(coords, n, buffer, ...)
}

#' @export
get_closest.download <- function(x, n, buffer, ...) {
  coords <- c(get_site(x)$long, get_site(x)$lat)
  get_closest(coords, n, buffer, ...)
}

R/param_check.R
#' @title Internal function to check passed parameters.
#'
#' @description Functions \code{\link{get_site}}, \code{\link{get_dataset}} and others pass parameters to \code{param_check}, which tells them if there's a problem.
#' @importFrom stringr str_detect
#' @importFrom dplyr filter
#' @param cl Contact ID is a numerical value associated with the Neotoma
#'    Contact table's numerical Contact ID.
#' @author Simon J. Goring \email{simon.j.goring@@gmail.com}
#' @return A list with two components:
#'
#'  \item{flag}{Returns a 0 if everything's fine, a 1 if there's a problem.}
#'  \item{message}{A list of error messages.}
#'
#' @references
#' Neotoma Project Website: http://www.neotomadb.org
#' API Reference:  http://api.neotomadb.org/doc/resources/contacts
#' @keywords internal misc
#' @export
param_check <- function(cl) {
  #  A long list of proper parameter formatting so that it's not stuck in all
  #  the basic functions.

  error <- list(flag = 0,
                message = list())

  if ('ageof' %in% names(cl)) {
    if (!cl$ageof %in% c('sample', 'taxon', 'dataset')) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- 'ageof parameter must be one of: sample, taxon or dataset'
    } else {
      if (any(c('taxonid', 'taxonname') %in% names(cl)) & !cl$ageof == 'taxon') {
        error$flag <- 1
        error$message[[length(error$message) + 1]] <- 'When taxonid or taxonname is invoked, ageof must be taxon'
      }
    }
    if (!any(c('ageyoung', 'ageold') %in% names(cl))) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- paste0('When ageof in invoked you also need to provide ',
                                                           'an age range using ageyounger or ageolder.')
    }
  }

  # Parameter check on ages:
  if (all(c('ageold', 'ageyoung') %in% names(cl))) {
    # If the user defines a minimum and maximum age, make sure that the
    # minimum is smaller than the max.
    if (cl$ageyoung > cl$ageold) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- paste0('ageyoung must be smaller than ageold, ',
                                                           'values were switched in call.')
    }
  }

  # Parameter check on altitudes.  This gets reused, we could turn it into a
  # higher level function to save reading lame code:
  if (all(c('altmin', 'altmax') %in% names(cl))) {
    # If the user defines a minimum and maximum altitude, make sure that the
    # minimum is smaller than the max.
    if (cl$altmin > cl$altmax) {
      error$flag = 1
      error$message[[length(error$message) + 1]] <- 'altmin must be smaller than altmax.'
    }
  }

  # Parameter check on contactid:
  if ('contactid' %in% names(cl)) {
    if (!is.numeric(cl$contactid)) {
      error$flag = 1
      error$message[[length(error$message) + 1]] <- 'The contactid must be numeric.'
    }
  }

  # Parameter check on contactname:
  if ('contactname' %in% names(cl)) {
    if (!is.character(cl$contactname)) {
      error$flag = 1
      error$message[[length(error$message) + 1]] <- 'The contactname must be a character string.'
    }
  }

  #  Parameter check on contactstatus:
  if ('contactstatus' %in% names(cl)) {
    if (!is.character(cl$contactstatus)) {
      error$flag = 1
      error$message[[length(error$message) + 1]] <- 'The contactstatus must be a character string.'
    } else {
      if (!cl$contactstatus %in% c('active', 'deceased', 'defunct',
                                   'extant', 'inactive', 'retired', 'unknown')) {
        error$flag = 1
        error$message[[length(error$message) + 1]] = paste0('status must be an accepted term.  ',
                                                            'Use get_table(\'ContactStatuses\')')
      }
    }
  }

  # Parameter check for the datasettype, make sure
  # it's one of the accepted types:
  if ('datasettype' %in% names(cl)) {
    settypes <- c('biochemistry', 'biomarker', 'charcoal', 'charcoal surface sample', 
                  'chironomid', 'cladocera', 'diatom', 'diatom surface sample', 
                  'diatom top-bottom', 'energy dispersive X-ray spectroscopy (EDS/EDX)', 
                  'geochemistry', 'geochronologic', 'insect', 'insect modern', 
                  'loss-on-ignition', 'macroinvertebrate', 'modern biochemistry', 
                  'modern plant biomarker', 'organic carbon', 'ostracode', 
                  'ostracode surface sample', 'paleomagnetic', 'physical sedimentology', 
                  'phytolith', 'plant macrofossil', 'pollen', 'pollen surface sample', 
                  'specimen stable isotope', 'stable isotope', 'testate amoebae', 
                  'testate amoebae surface sample', 'vertebrate fauna', 'water chemistry', 
                  'X-ray diffraction (XRD)', 'X-ray fluorescence (XRF)')
    
    set <- pmatch(cl$datasettype, settypes, nomatch = NA)
    if (is.na(set)) {
      warning(paste0('datasettype is not within the known set of records:: \n* ',paste0(settypes, collapse = '\n* '), ''))
    }
  }
  
  # Parameter check on familyname:
  if ('familyname' %in% names(cl)) {
    if (!is.character(cl$familyname)) {
      error$flag = 1
      error$message[[length(error$message) + 1]] <- 'The familyname must be a character string.'
    }
  }

  #  Test the geographic identification against the Geopolitical name table.
  if ('gpid' %in% names(cl)) {
    if (is.character(cl$gpid)) {
      
      good_gp <- "province|state|territor|country"
      places <- paste0("^(", paste0(cl$gpid, collapse = "|"), ")$")
      
      gp_test <- gp.table %>% 
        dplyr::filter(stringr::str_detect(gp.table$GeoPoliticalUnit, good_gp) &
               stringr::str_detect(gp.table$GeoPoliticalName, places))
      
      if (nrow(gp_test) == 0) {
        error$flag <- 1
        message <- paste0('Cannot find a match for the gpid provided.')
        error$message[[length(error$message) + 1]] <- message
      }
      if (nrow(gp_test) > length(cl$gpid)) {
        error$flag <- 1
        message <- paste0(length(cl$gpid), " geopolitical name(s) were passed, but Neotoma matches ", nrow(gp_test), " locations.  Please check location names against the neotoma GeoPoliticalUnits table using get_table('GeoPoliticalUnits')")
        error$message[[length(error$message) + 1]] <- message
      }
      cl$gpid <- gp_test$GeoPoliticalID
    } else {
      if (!is.numeric(cl$gpid)) {
        error$flag <- 1
        error$message[[length(error$message) + 1]] <- 'The gpid must be either a character string or an integer.'
      }
    }
  }

  # Parameter check on 'loc', ought to be a comma separated list of
  # lonW, latS, lonE, latN when it is passed out, but it's probably
  # better to pass in a vector.  Although it might be better to associate
  # it with a spatial object existing in R like an extent or bbox.
  if ('loc' %in% names(cl)) {
    cl$loc <- eval(cl$loc)

    if (class(cl$loc) == 'numeric' & length(cl$loc == 4)) {

      # The latitudes must be from -90 to 90
      # The longitudes must be from -180 to 180
      if (all(cl$loc[c(2,4)] >= -90 & cl$loc[c(2,4)] <= 90) &&
          all(cl$loc[c(1,3)] >= -180 & cl$loc[c(2,4)] <= 180)) {
        cl$loc <- paste(cl$loc, collapse = ',')
      } else {
        error$flag <- 1
        error$message[[length(error$message) + 1]] <- paste0('loc must be in the form c(lonW, latS, lonE, latN).',
                    '\nLongitudes from -180 to 180, latitudes from -90 to 90.')
      }
    } else {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- 'The loc must be a numeric vector: lonW, latS, lonE, latN.\n'
    }
  }

  #  Test the PI identification.
  if ('piid' %in% names(cl)) {
    # piid must be the numeric PI id number in the Neotoma database.
    if (!is.numeric(cl$piid)) {
      error$flag = 1
      error$message[[length(error$message) + 1]] = 'piid must be a numeric value.'
    }
  }

  # Parameter check on siteid:
  if ('siteid' %in% names(cl)) {
    if (!is.numeric(cl$siteid)) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- 'siteid must be numeric.'
    }
  }

  if ('taxonname' %in% names(cl)) {
    if (!class(cl$taxonname) == 'character') {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- 'The taxonname must be a character.'
    }
  }
  
  # Parameter check on author:
  if ('search' %in% names(cl)) {
    if (!is.character(cl$search)) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- 'The search string must be a character string.'
    }
  }
  
  if ('year' %in% names(cl)) {
    if (!is.numeric(cl$year)) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- 'The year used must be numeric.'
    }
  }
  
  if ('pubtype' %in% names(cl)) {
    if (!is.character(cl$pubtype)) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- paste0('The pubtype must be a character string. Use get.table',
                  '(\'PublicationTypes\') to find acceptable tables.')
    }
  }
  
  # Parameter check on pubid:
  if ('pubid' %in% names(cl)) {
    if (!is.numeric(cl$pubid)) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- 'The pubid must be numeric.'
    }
  }
  
  # Parameter check on contactid:
  if ('contactid' %in% names(cl)) {
    if (!is.numeric(cl$contactid)) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- 'The contactid must be numeric.'
    }
  }
  
  # Parameter check on datasetid:
  if ('datasetid' %in% names(cl)) {
    if (!is.numeric(cl$datasetid)) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- 'The datasetid must be numeric.'
    }
  }
  
  # Parameter check on author:
  if ('author' %in% names(cl)) {
    if (!is.character(cl$author)) {
      error$flag <- 1
      error$message[[length(error$message) + 1]] <- 'The author name must be a character string.'
    }
  }
  
  list(cl,error)

}
